### PR TITLE
DALI support for Tasmota

### PR DESCRIPTION
**Pull request by eeak**
## Description:

**Related issue (if applicable):** fixes #<Tasmota issue number goes here>

## Checklist:
  - [x ] The pull request is done against the latest development branch
  - [x ] Only relevant files were touched
  - [x ] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x ] The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [ x] The code change is tested and works with Tasmota core ESP32 V.2.0.5
  - [x ] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

### CODE REVIEW

**Review by barbudor — COMMENTED** [review]
This is lacking explanations
What is it intended to offer ?
Does it makes a Tasmota a DALI device capable of being controlled through DALI protocol ?
Or is it a gateway to control DALI devices from MQTT ?

There is quite some un-used code in comments which reduce the readability.
If you intend to enhance the driver later, could you for the moment remove the un-used code and add it later ?

I put a few comments also in the code as it is using quite some RAM even when the driver is not in use. You should use dynamically allocated memory lke it is done in recent drivers.

Preferably, my_user_config.h should include all concerned defines but in comment so people know what is required to enable it and copy the line to their user_config_override.h

Thanks

File: tasmota/include/tasmota_template.h [hunk 1]
```diff
@@ -198,6 +198,9 @@ enum UserSelectablePins {
   GPIO_ADE7953_RST,                    // ADE7953 Reset
   GPIO_NRG_MBS_TX, GPIO_NRG_MBS_RX,    // Generic Energy Modbus device
   GPIO_ADE7953_CS,                     // ADE7953 SPI Chip Select
+#ifdef ESP32
```
Comment (s-hadinger) [inline tasmota/include/tasmota_template.h:201]: The numbers should not depend on compilation option. Please remove #ifdef

**Comment by eeak** [comment]
> Does it makes a Tasmota a DALI device capable of being controlled through DALI protocol ?
> Or is it a gateway to control DALI devices from MQTT ?

This is initial DALI support. Now tasmota can control DALI devices via MQTT and listen the DALI bus to build an actuator.

> There is quite some un-used code in comments which reduce the readability.
> If you intend to enhance the driver later, could you for the moment remove the un-used code and add it later ?

I will do

> Preferably, my_user_config.h should include all concerned defines but in comment so people know what is required to enable it and copy the line to their user_config_override.h

I will do too

**Comment by barbudor** [comment]
> This is initial DALI support. Now tasmota can control DALI devices via MQTT and listen the DALI bus to build an actuator.

So it's more like the idea of a gateway between MQTT and DALI lights.
So I would expect the driver to expose some virtual relay or virtual dimmers that could be controlled through standard Tasmota commands (POWER, DIMMER...) and translate the commands into DALI commands
The lights controlled through Tasmota would then appear as any tasmota light within any home automation system

File: tasmota/include/tasmota_template.h (hunk 1 above)
Comment (eeak) [reply tasmota/include/tasmota_template.h:201]: But I see several similar ones above

File: tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+
+const char HTTP_BTN_MENU_DALI[] PROGMEM =
+  "<p><form action='" WEB_HANDLE_DALI "' method='get'><button>" D_CONFIGURE_DALI "</button></form></p>";
+
+#endif // USE_WEBSERVER
+
+
+
+#define DALI_TOPIC "DALI"
+static char tmp[120];
```
[the remark this answers is not in the document]
Comment (eeak) [reply tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino:450]: remove this unused array

File: tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+
+const char S_JSON_DALI_COMMAND_NVALUE[] PROGMEM = "{\"" D_NAME_DALI "\":{\"%s\":%d}}";
+const char kDALI_Commands[] PROGMEM  = D_CMND_DALI_POWER "|" D_CMND_DALI_DIMMER;
+
+enum DALI_Commands {         // commands for Console
+  CMND_DALI_PWR,
+  CMND_DALI_DIM,
+};
+
+/* Private variables ---------------------------------------------------------*/
```
[the remark this answers is not in the document]
Comment (Staars) [reply tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino:59]: These variables are totally global in Tasmota and IMHO something like ˋflagˋ is a „dangerous“ name. You could move them into a struct or use prefixes.

File: tasmota/include/tasmota_template.h (hunk 1 above)
Comment (s-hadinger) [reply tasmota/include/tasmota_template.h:201]: Oh yes. Sorry my bad. I misread as a Dali I def. Please ignore my comment

File: tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino (hunk 3 above)
[the remark this answers is not in the document]
Comment (eeak) [reply tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino:56]: Packed to struct

File: tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+
+void DaliPreInit() {
+    if (!PinUsed(GPIO_DALI_TX) || !PinUsed(GPIO_DALI_RX)) { return; }
+    AddLog(LOG_LEVEL_INFO, PSTR("DLI: Init - RX-pin: %d, TX-pin: %d"), Pin(GPIO_DALI_RX), Pin(GPIO_DALI_TX));
+    // pinMode(LED, OUTPUT);
+	pinMode(Pin(GPIO_DALI_TX), OUTPUT);
+	digitalWrite(Pin(GPIO_DALI_TX), HIGH);
+	pinMode(Pin(GPIO_DALI_RX), INPUT);
+
+	DALI_timer = timerBegin(0, 13, true);
```
[the remark this answers is not in the document]
Comment (eeak) [reply tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino:393]: Define in my_user_config.h ?

[the remark this answers is not in the document]
Comment (barbudor) [reply tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino:393]: A good way is to have a commented defined in my_user_config.h

// #define DALI_TIMER 0  // default hw timer used by Dali driver. Can be overridden in user_config_override.h

And in your code

#ifndef DALI_TIMER 
#define DALI_TIMER 0
#endif

Anyone wanting to change can just do a
#define DALI_TIMER 1
in his user_config_override.h

[the remark this answers is not in the document]
Comment (eeak) [reply tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino]: Thank you!

File: tasmota/my_user_config.h [hunk 5]
```diff
@@ -1034,6 +1034,11 @@
 
 #define USE_ESP32_SENSORS                        // Add support for ESP32 temperature and optional hall effect sensor
 
+#define USE_DALI                              // Add support for DALI
```
Comment (barbudor) [inline tasmota/my_user_config.h:1037]: Please comment that line, it should not be enabled by default
Thanks

File: tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino [hunk 6]
```diff
@@ -0,0 +1,582 @@
...
+    uint8_t flag;        // DALI status flag
+    uint8_t bit_count;   // nr of rec/send bits
+    uint16_t tick_count; // nr of ticks of the timer
+    bool former_val;     // bit value in previous tick of timer
+
+    bool present = false;// DALI initialized
+} Dali;
+
+hw_timer_t *DALI_timer = NULL;
+
```
Comment (barbudor) [inline tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino:74]: Move hw timer into the struct
Make Dali a pointer initialized as nullptr
```C++
struct DALI {
 ....
  hw_timer_t *timer;
} *Dali = nullptr;
```

File: tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino [hunk 7]
```diff
@@ -0,0 +1,582 @@
...
+    }
+}
+
+/**
+* @brief  This function enable data transfer start interrupt.
+* @param  None
+* @retval None
+*/
+void enableDaliRxInterrupt() {
+    Dali.flag = DALI_NO_ACTION;
```
Comment (barbudor) [inline tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino:102]: Change all `Dali.` in `Dali->`

File: tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino [hunk 8]
```diff
@@ -0,0 +1,582 @@
...
+
+void DaliPreInit() {
+    if (!PinUsed(GPIO_DALI_TX) || !PinUsed(GPIO_DALI_RX)) { return; }
+    AddLog(LOG_LEVEL_INFO, PSTR("DLI: Init - RX-pin: %d, TX-pin: %d"), Pin(GPIO_DALI_RX), Pin(GPIO_DALI_TX));
+    // pinMode(LED, OUTPUT);
+	pinMode(Pin(GPIO_DALI_TX), OUTPUT);
+	digitalWrite(Pin(GPIO_DALI_TX), HIGH);
+	pinMode(Pin(GPIO_DALI_RX), INPUT);
+
+	DALI_timer = timerBegin(DALI_TIMER, 13, true);
```
Comment (barbudor) [inline tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino:384]:
```C++
Dali = (DALI*)calloc(1,sizeof(DALI));
if (!Dali) {
  AddLog(LOG_LEVEL_INFO, PSTR("DLI: Memory allocation error"));
  return;
}
Dali->timer = timerBegin(DALI_TIMER, 13, true);
```

File: tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino [hunk 9]
```diff
@@ -0,0 +1,582 @@
...
+
+bool Xdrv89(uint8_t function)
+{
+    bool result = false;
+
+    if (FUNC_INIT == function)
+    {
+        DaliPreInit();
+    }
+    else if (Dali.present)
```
Comment (barbudor) [inline tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino:566]: You don't need `present` anymore. `if(Dali)` is now equivalent.
```C++
if (Dali)
{
  switch(function)
  {
  ...
   }
}
```

File: tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino [hunk 10]
```diff
@@ -0,0 +1,582 @@
...
+	digitalWrite(Pin(GPIO_DALI_TX), HIGH);
+	pinMode(Pin(GPIO_DALI_RX), INPUT);
+
+	DALI_timer = timerBegin(DALI_TIMER, 13, true);
+    timerAttachInterrupt(DALI_timer, &DALI_Tick_Handler, true);
+    timerAlarmWrite(DALI_timer, 641, true);
+    
+    attachInterrupt(Pin(GPIO_DALI_RX), receiveDaliData, FALLING);
+	enableDaliRxInterrupt();
+    Dali.present = true;
```
Comment (barbudor) [inline tasmota/tasmota_xdrv_driver/xdrv_89_esp32_dali.ino:390]: You don't need that `present` anymore
Testing `if(Dali)` does the same now

**Review by barbudor — COMMENTED** [review]
Please see my new comments
This is what I meant about using dynamically allocated private data.
That is avoiding using RAM is the driver is compiled-in but not used.

**Comment by barbudor** [comment]
👍

File: tasmota/include/tasmota_template.h (hunk 1 above)
Comment (arendst) [inline tasmota/include/tasmota_template.h:201]: I understand the idea but let's keep the list in sync between both architrctures to make documenting easier.

So pls remove the test for ESP32 here

File: tasmota/include/tasmota_template.h [hunk 11]
```diff
@@ -443,6 +446,9 @@ const char kSensorNames[] PROGMEM =
   D_SENSOR_ADE7953_RST "|"
   D_SENSOR_NRG_MBS_TX "|" D_SENSOR_NRG_MBS_RX "|"
   D_SENSOR_ADE7953_CS "|"
+#ifdef ESP32
+  D_SENSOR_DALI_RX "|" D_SENSOR_DALI_TX "|"
```
Comment (arendst) [inline tasmota/include/tasmota_template.h:450]: Ditto

File: tasmota/include/tasmota_template.h [hunk 12]
```diff
@@ -527,6 +533,11 @@ const uint16_t kGpioNiceList[] PROGMEM = {
  * Protocol specifics
 \*-------------------------------------------------------------------------------------------*/
 
+#if defined(USE_DALI) && defined(ESP32)
```
Comment (arendst) [inline tasmota/include/tasmota_template.h:536]: This is OK

File: tasmota/language/af_AF.h [hunk 13]
```diff
@@ -511,6 +511,13 @@
 #define D_ZIGBEE_EEPROM_FOUND_AT_ADDRESS "ZBBridge EEPROM gevind by adres"
 #define D_ZIGBEE_RANDOMIZING_ZBCONFIG "Ewekansige Zigbee parameters, kyk asseblief met 'ZbConfig'"
 
+// xdrv_89_dali.ino
+#define D_SENSOR_DALI_RX                  "Dali RX"
+#define D_SENSOR_DALI_TX                  "Dali TX"
+#define D_CONFIGURE_DALI                  "Config DALI"
+#define D_CMND_DALI_POWER                 "power"
```
Comment (arendst) [inline tasmota/language/af_AF.h:518]: D_CMND_xx always needs to be english and hence are defined in include/i18n.h

Pls remove from all language files

File: tasmota/include/tasmota_template.h (hunk 11 above)
Comment (eeak) [reply tasmota/include/tasmota_template.h:447]: performed

File: tasmota/include/tasmota_template.h (hunk 1 above)
Comment (eeak) [reply tasmota/include/tasmota_template.h]: performed

File: tasmota/language/af_AF.h (hunk 13 above)
Comment (eeak) [reply tasmota/language/af_AF.h]: moved

**Comment by Noschvie** [comment]
Is there a connecting diagram ? Is the ESP32 connected direct to the DALI pins (DA 1 and DA 2) of a LED converter / controller ?

Can this be a replacement for a "Dali - Zigbee converter" - Sonoff Zigbee Pro running Tasmota ? Discussions #16935

**Comment by eeak** [comment]
> Is there a connecting diagram ? Is the ESP32 connected direct to the DALI pins (DA 1 and DA 2) of a LED converter / controller ?

The Dali bus is about 16 volts. Of course you cannot connect directly, Only through the optocouples. I use like this:
photo_2022-10-31_09-48-43

**Comment by Noschvie** [comment]
> 
> The Dali bus is about 16 volts. Of course you cannot connect directly, Only through the optocouples. I use like this: photo_2022-10-31_09-48-43

Ok, understand, maybe you have a Gerber file (PCB) for a test run ?

**Comment by eeak** [comment]
[DALI driver_2022-10-31.zip](https://github.com/arendst/Tasmota/files/9903628/DALI.driver_2022-10-31.zip)
Sketched it out quickly. Can use SMD or DIP elements.

**Comment by Noschvie** [comment]
Thanks @eeak
Found this [dali-2-click](https://www.mikroe.com/dali-2-click)
is it supported to be connected to a ESP32 running Tasmota ?

**Comment by barbudor** [comment]
I think it would be nice to start a dedicated Doc page with build instructions, tested hardware and usage exemples

**Comment by eeak** [comment]
> Thanks @eeak Found this [dali-2-click](https://www.mikroe.com/dali-2-click) is it supported to be connected to a ESP32 running Tasmota ?

Yes. But probably have to adjust the invert:

**Comment by Noschvie** [comment]
Will order this [dali-2-click](https://www.mikroe.com/dali-2-click) and report back.

**Comment by Noschvie** [comment]
@eeak : how have the [dali-2-click](https://www.mikroe.com/dali-2-click) to be connected with the ESP32 ?
3.3V -> ESP32 3.3V
GND -> GND
RX -> ESP32 RX
TX -> ESP32 TX

I'm using an ESP32 "Wemos C3 mini V2.1.0"

**Comment by Noschvie** [comment]
Did some tests but didn't get any response from the DALI device. Don't have the skills and the environment to debug this.

**Comment by eeak** [comment]
> Did some tests but didn't get any response from the DALI device. Don't have the skills and the environment to debug this.

I hope you didn't forget about dali bus power supply ?

**Comment by Noschvie** [comment]
The dali LED converter is connected to 220V.
Do you mean that DALI RX and TX have to be connected to a "Dali power supply" / 16 VDC too ?

**Comment by eeak** [comment]
> The dali LED converter is connected to 220V. Do you mean that DALI RX and TX have to be connected to a "Dali power supply" / 16 VDC too ?

изображение

**Comment by modelD-svg** [comment]
> Thanks @eeak Found this [dali-2-click](https://www.mikroe.com/dali-2-click) is it supported to be connected to a ESP32 running Tasmota ?

I am new to dimming and dali in general & interested in buying a dali-capable fixture. Glad to see it being supported!

Would you mind sharing your setup with the dali2click in more detail, if you get it to work? Would probably save me lots of time and fiddling. Thanks!

**Comment by Noschvie** [comment]
Will get the Dali power supply next week and shall be able to test the [dali-2-click](https://www.mikroe.com/dali-2-click) device.

**Comment by marcobrianza** [comment]
hello, I have made a DALI shield for Wemos board and will try it soon with the new Tasmota release
https://oshwlab.com/marco.brianza/dalino

**Comment by eeak** [comment]
> hello, I have made a DALI shield for Wemos board and will try it soon with the new Tasmota release https://oshwlab.com/marco.brianza/dalino

Thank you, Marco. But note - only works on esp32 core (Wemos C3 and S2).

**Comment by marcobrianza** [comment]
ok, I have both, If you want an assembled board I can send you one to test write me at marcobrianza at gmail

**Comment by kettenbach-it** [comment]
Can someone recommend a cheap dali power supply for only one device (Osram Arktika LED lamp)?

I'v seen the DALI patch makde it to 12.3.1. Does ist work only with ESP32 or will ESP12 (8266) as well?

**Comment by eeak** [comment]
> Can someone recommend a cheap dali power supply for only one device (Osram Arktika LED lamp)?

Dimmable ? AC230 or DC12 volts ?

> I'v seen the DALI patch makde it to 12.3.1. Does ist work only with ESP32 or will ESP12 (8266) as well?

Due to the use of a hardware timer, it only works on the esp32 core

**Comment by kettenbach-it** [comment]
> > Can someone recommend a cheap dali power supply for only one device (Osram Arktika LED lamp)?
> 
> Dimmable ? AC230 or DC12 volts ?

Dimmable with AC230

**Comment by eeak** [comment]
> AFAIK I need to power (only) the dali bus.

I use this on several objects.
[https://u.eeak.biz/euxw9](url)

**Comment by kettenbach-it** [comment]
Thanks!

One more thing: is the DALI support in the release build included or do I need to compile myself? How's it called (in the dropdown for the PIN)?

**Comment by kettenbach-it** [comment]
I'm answering my own question: No, it's not.

It needs to enabled here: https://github.com/arendst/Tasmota/blob/e8edf976dacf1e636a4119268280c983a595c6c3/tasmota/my_user_config.h#L1058

**Comment by kettenbach-it** [comment]
To whom it may concern: this is how it should look like:

image

Waiting for the power supply to arrive and then I'll build a HAT to adapt to the Dali bus

**Comment by sergios100** [comment]
Hi!
I compiled the ESP32 Version with enabled Dali Support and inverted Signals:

2022-12-28_21h56_49

and selected Dali RX and Dali TX in the Module Configuration:
2022-12-28_21h44_09
I hooked up a DALI click Board from Mikroelektronika to these two pins (plus 3.3V and GND). On the other side (the DALI-Side of the click Board) is a Osram DALI FL-Fixture and a 12V Powersupply.
But how do I control the Light-Fixture?
By Commands? Which commands? Nothing seems to work:

2022-12-28_21h49_36

Or should it be controllable from the Main Menu? There is also nothing:

2022-12-28_21h51_24

Any hints appreciated!

Cheers sergio

**Comment by sergios100** [comment]
Sorry to bother again, but I tried different commands in the console, just got errors back.
grafik

Could somebody just write me in plain text what to enter in the console to send an "Power ON" command to the DALI driver?

Thanks!
Sergio

**Comment by kettenbach-it** [comment]
I also can not understand which commands these should be @eeak

**Comment by sfromis** [comment]
Those upper-case names with underscore are symbolic names in the C++ code, not names. From what I can see in the code, the command names should be `DaliPower` and `DaliDim`, the constant values prefixed with `D_NAME_DALI` having the value of `DALI`.

**Comment by kettenbach-it** [comment]
Can confirm `DaliPower`(without knowing if it really works, since I do not yet have the bus-coupler). But can't conform `DaliDim`:

image

**Comment by sfromis** [comment]
I do not change my statement about `DaliDim` as a command name matching the source code, but that does not mean that the `DaliDim` command is actually implemented. For whatever weird reason, it will parse the command name, but after that it has no processing of it, causing it to not be picked up after all.

**Comment by sergios100** [comment]
Hi Guys!

Thanks for the support! 
I got this far: `DaliPower 0` to `DaliPower 254` sets the brightness of the fixture from 0 = OFF to 254 = FULL-Brightness!
(Values in Between can be used to dim the fixture. Ex. DaliPower 128 sets the fixture to about half brightness)

To whom it might be of interest:
I am using a MikroElektronica "Dali Click board" [https://www.mikroe.com/dali-click](url) 
Wiring is done as follows:

ESP 32             |          Dali Click
_______ | _______
 3.3V                |           3.3V
GND               |           GND
DALI TX          |            TX
DALI RX          |            INT

Software has been compiled with these Settings:

`#define USE_DALI  `
   `  #define DALI_IN_INVERT  0  `
  `  #define DALI_OUT_INVERT  0  `
   `  #define DALI_TIMER 0  `

To power the bus, I used the weakest Power Supply I could find (12V, 600mA, "Wal Wart") and added a 300 Ohm Resistor in series. This is to reduce the amps running through the Mosfet on the Click Board when it "pulls down the bus" to send data.

As mentioned above by sfromis the command `DaliDim` does not seem to be implemented (yet).
Also recieving from the bus seems not yet to be implemented, as I have hooked up an other "Master" on the bus (Nexentro Zigbee -> DALI Interface), I see the orange LED on the "Dali Click" blink when the Nexentro Master is sending data to the bus, but I don't see anything in the console.

I guess the DALI implementation has just begun?
Would be realy cool If at some point DALI Fixtures could be controlled from the GUI as well, perhaps looking something like this:

2023-01-07_19h10_58

And perhaps the possibility to send "raw data" to the DALI bus in this form `YAAA AAAS xxxx xxxx` like described here:
[https://infosys.beckhoff.com/english.php?content=../content/1033/kl6811/2214047371.html&id=](url)

Cheers! sergio

**Comment by eeak** [comment]
Hi all.
I have plans to develop DALI support, but unfortunately I have little free time.
You can sketch your ideas about DALI development here.
In the near future, I have to do the initialization of DALI devices.

For urgent questions, you can contact telegram: [https://t.me/tasmota_ru](https://t.me/tasmota_ru)

**Comment by ByteSizedMarius** [comment]
Hello, thanks to @eeak, @sergios100 and everyone else in this thread, I got this working today using an ESP-32S, the Dali-2-Click board and the AliExpress Dali Power-Supply eeak recommended (eeaks link is dead, here it is again: aliexpress.com/item/4000016930121.html -- Miboxer DALI Bus Power). Sharing some information here, maybe it will help someone in a similar situation.

For whatever reason, I was unable to power the dali-2-click through 2 seperate Wemos C3 mini V2.1.0 boards. Whenever I connected the d2c it would get power, but the wemos would shut down (tried multiple different power supplies, because I suspected power to be the issue). It worked first try with the 32S.

I wanted to use 2 buttons to control the fixture, one for toggle power and one for dimming. I quickly implemented this using berry (excuse messy code, this was my first time using berry).

```berry
var dalidimstate = 0
var dalipowerstate = 0
tasmota.cmd("DaliPower 100")

def dali_toggle_pwr()
    if dalipowerstate == 0
        if dalidimstate == 1
             tasmota.cmd("DaliPower 100", true)
        else
             tasmota.cmd("DaliPower 254", true)
        end
        dalipowerstate = 1
    else
        tasmota.cmd("DaliPower 0", true)
        dalipowerstate = 0
    end

    tasmota.resp_cmnd_done()
end


def dali_toggle_dim()
    if dalidimstate != 1
        tasmota.cmd("DaliPower 100", true)
        dalidimstate = 1
    else
        tasmota.cmd("DaliPower 254", true)
        dalidimstate = 0
    end

    dalipowerstate = 1
    tasmota.resp_cmnd_done()
end

tasmota.add_cmd("DaliToggleDim", dali_toggle_dim)
tasmota.add_cmd("DaliTogglePwr", dali_toggle_pwr)
```

This script adds the commands `DaliToggleDim` and `DaliTogglePwr`. To execute the script on boot, create a file called `autoexec.be` using the tasmota-filesystem and put the code in there.

The first dali command after boot seems to go missing for me, which is why i added the `DaliPower 100`. Not entirely sure if this is an issue with my fixture or the implementation.

**Comment by Vall2kas** [comment]
And perhaps the possibility to send "raw data" to the DALI bus in this form YAAA AAAS xxxx xxxx like described here ...
And why not: DALI bridge, mqtt->DALI bus ?

**Pull request merged**